photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/RandomSearch.scala

```diff
@@ -117,39 +105,25 @@ class RandomSearch[T](
 
     require(n > 0, "The number of results must be greater than zero.")
 
-    val candidate = drawCandidates(1)(0,::).t
+    val candidate = drawCandidates(1)(0, ::).t
 
     // Make values discrete as specified
-    discreteParams.foreach { index =>
-      candidate(index) = round(candidate(index))
-    }
+    discretizeCandidate(candidate, discreteParams)
 
     val (_, model) = evaluationFunction(candidate)
+    val initialObservation = evaluationFunction.convertObservations(Seq(model))
 
-    Seq(model) ++ (if (n == 1) Seq() else findWithPriors(n - 1, convertObservations(Seq(model)), priorObservations))
+    Seq(model) ++ (if (n == 1) Seq() else findWithPriors(n - 1, initialObservation, priorObservations))
   }
 
+
   /**
    * Searches and returns n points in the space.
    *
    * @param n The number of points to find
    * @return The found points
    */
-  def find(n: Int): Seq[T] = findWithPriorObservations(n, Seq())
-
-  /**
-   * Vectorize a [[Seq]] of prior observations.
-   *
-   * @param observations Prior observations in estimator output form
-   * @return Prior observations as tuples of (vector representation of the original estimator output, evaluated value)
-   */
-  def convertObservations(observations: Seq[T]): Seq[(DenseVector[Double], Double)] =
-    observations.map { observation =>
-      val candidate = evaluationFunction.vectorizeParams(observation)
-      val value = evaluationFunction.getEvaluationValue(observation)
-
-      (candidate, value)
-    }
+  def find(n: Int): Seq[T] = findWithPriors(n, Seq(), Seq())
```

Comment:
I think what @joshvfleming was saying could be used here.

We can call find with the observations and priorObservations. If observations is empty we call findWithPriorObservations else we call findWithPriors. In that case, we can remove the if from the findWithPriors and the code becomes cleaner. 

Otherwise now, from find we are calling findWithPriors with empty observations which is then calling findWithPriorObservations.

Reply:
I suggest we still add the if-statement in findWithPrior anyway for the safety of the code. After all,
findWithPrior(n, Seq(), Seq()) is a legal call but it threw an error in the previous code.

Reply:
And, yes, we can just change `find` back to:
`def find(n: Int): Seq[T] = findWithPriorObservations(n, Seq())`

Reply:
@cmjiang Let's move the `if` statement to the `find` function, and add an invariant check to `findWithPrior`, e.g.:

```scala
  require(observations.nonEmpty, "There must be at least one observation")
```

Reply:
Actually, thinking about it more, we don't need an `if` statement. The `find` function could have the implementation that assumes no prior information, and `findWithPrior` would assume prior information as an input invariant, as before.

Reply:
@cmjiang Let's punt on this and go back to the way it was before (but with the `require` invariant check -- that was the actual bug here). I think we need to re-think how this is organized a little to make it "right".

Reply:
+1 on ^